### DISCUSSION TITLE
Reconsider composition state variables for reactors, flames, and other systems

### DISCUSSION THREAD

**Discussion by speth** [Cantera/enhancements#91] [Ideas]
**Motivation**

Cantera's governing equations for reactors and flames currently use mass fractions as the state variables for the composition. This approach leads to fairly simple equations for the most common cases, but introduces some thorny mathematical issues due to the constraint that the sum of the mass fractions should be 1 that make it difficult to solve certain problems. While we have workarounds in place for some of these issues, these workarounds are often fragile. Some specific examples:

- Calculating reasonable Jacobians requires leaving the sum(Y) = 1 constraint unsatisfied. In this case, there is no fully consistent way to calculate mole fractions, or any quantities which are computed from mole fractions. This is a concern for computing mixture averaged transport fluxes, proportional to the mole fraction gradients, and for any non-ideal thermo model where properties are calculated based on mole fractions.

- Without the sum(Y) = 1 constraint, steady-state solvers are difficult to use because (depending on the situation), they can find the "solution" of setting all mass fractions to zero. This is particularly a problem for steady-state surface coverage problems.

**Possible Solutions**

These issues could be resolved by using concentrations or species masses as the species state variables instead.

**References**

Cantera/cantera#390

**Comment by kyleniemeyer** [comment]
cc @anthony-walker

**Comment by anthony-walker** [comment]
This would definitely coincide well with my [work](https://github.com/Cantera/cantera/pull/951).

**Comment by ischoegl** [comment]
This approach could also extend to handling of `Mixture`s, where mole/mass fractions do not uniquely define the overall composition.

**Comment by speth** [comment]
@ischoegl - That's a good point. Using the mass or moles of each species in each phase would avoid having to separately track the mass of each phase as a separate quantity. I'm wondering now if there's a benefit to applying this idea at the `ThermoPhase` level, which would effectively make a `ThermoPhase` object the equivalent of the Python `Quantity` class (phase + ammount), and provide a very simple implementation of `Mixture`.

@anthony-walker - could you summarize anything you've found that relates to this, either in terms of the issues with using mass/mole fractions, or any challenges you've found in using masses or concentrations?

**Comment by kyleniemeyer** [comment]
Some useful references that discuss this issue: 
- https://doi.org/10.1016/j.combustflame.2018.03.017
- https://doi.org/10.1016/S0010-2180(01)00352-2

**Comment by ischoegl** [comment]
@speth ... I'd be very much in favor of handling this at the `ThermoPhase` level. But I believe it may be possible to get away without having to handle amount by using concentrations (kmol/m^3)? I believe an equation of state should be able to handle the volume portion based on pressure and temperature, although I am not familiar enough with non-ideal gases. This probably needs further thought and I am sure that other thorny issues exist that I am currently unaware of.

**Comment by anthony-walker** [comment]
@speth The issues I have had so far have been fairly specific to my work and I don't have a lot of experience with Cantera's other functionalities that may be affected by such a change. 

The first and foremost is that the Jacobian of a constant pressure mass-fraction based system is fully dense and my work relies on leveraging sparse solvers. So, to precondition the current system a hack is in place to copy and convert the system to mole fractions. Something like [T,m,X_1..X_n], solve the linear system for an output vector used by Sparse GMRES (Sundials) with a mole fraction Jacobian preconditioner, and then convert the output vector to a state consistent to the current state that cantera can handle. I believe it is possible to make this same preconditioner based in concentration and not mole fractions as well so it could be consistent and not need any conversions.

The concentrations are necessary for all of the calculations used to build this preconditioner, so a conversion would also simplify that process as well. A hack or two may also be in place for this if I remember correctly so that I can obtain the appropriate information from the `Thermo` and `Kinetics` objects.  I will have to look further into this for more details though.

I will append more to this as I think of it.

**Comment by speth** [comment]
Superseded by #154 - closing to avoid splitting any discussion.
